src/rx5808-pro-diversity/receiver.cpp

```diff
@@ -0,0 +1,227 @@
+#include <Arduino.h>
+#include <avr/pgmspace.h>
+
+#include "settings.h"
+#include "settings_eeprom.h"
+#include "receiver.h"
+#include "receiver_spi.h"
+#include "channels.h"
+
+#include "timer.h"
+
+static void updateRssiLimits();
+static void writeSerialData();
+
+
+namespace Receiver {
+    ReceiverId activeReceiver = ReceiverId::A;
+    uint8_t activeChannel = 0;
+
+    uint8_t rssiA = 0;
+    uint16_t rssiARaw = 0;
+    uint8_t rssiALast[RECEIVER_LAST_DATA_SIZE] = { 0 };
+    #ifdef USE_DIVERSITY
+        uint8_t rssiB = 0;
+        uint16_t rssiBRaw = 0;
+        uint8_t rssiBLast[RECEIVER_LAST_DATA_SIZE] = { 0 };
+
+        ReceiverId diversityTargetReceiver = activeReceiver;
+        Timer diversityHysteresisTimer = Timer(DIVERSITY_HYSTERESIS_PERIOD);
+    #endif
+
+    uint32_t lastChannelSwitchTime = 0;
+    static Timer rssiLogTimer = Timer(RECEIVER_LAST_DELAY);
+    #ifdef USE_SERIAL_OUT
+        static Timer serialLogTimer = Timer(25);
+    #endif
+
+
+    void setChannel(uint8_t channel)
+    {
+        ReceiverSpi::setSynthRegisterB(Channels::getSynthRegisterB(channel));
+
+        lastChannelSwitchTime = millis();
+        activeChannel = channel;
+    }
+
+    void setActiveReceiver(ReceiverId receiver) {
+        #ifdef USE_DIVERSITY
+            #ifdef USE_DIVERSITY_FAST_SWITCHING
+                uint8_t targetPin, disablePin;
+                if (receiver == ReceiverId::A) {
+                    targetPin = PIN_LED_A;
+                    disablePin = PIN_LED_B;
+                } else {
+                    targetPin = PIN_LED_B;
+                    disablePin = PIN_LED_A;
+                }
+
+                uint8_t port = digitalPinToPort(targetPin);
+                uint8_t targetBit = digitalPinToBitMask(targetPin);
+                uint8_t disablebit = digitalPinToBitMask(disablePin);
+                volatile uint8_t *out = portOutputRegister(port);
+
+                *out = (*out | targetBit) & ~disablebit;
+            #else
+                digitalWrite(PIN_LED_A, receiver == ReceiverId::A);
+                digitalWrite(PIN_LED_B, receiver == ReceiverId::B);
+            #endif
+        #else
+            digitalWrite(PIN_LED_A, HIGH);
+        #endif
+
+        activeReceiver = receiver;
+    }
+
+    //
+    // Blocks until MIN_TUNE_TIME has been reached since last channel switch.
+    //
+    void waitForStableRssi() {
+        uint16_t timeSinceChannelSwitch = millis() - lastChannelSwitchTime;
+        if (timeSinceChannelSwitch < MIN_TUNE_TIME) {
+            delay(MIN_TUNE_TIME - timeSinceChannelSwitch);
+        }
+    }
+
+    uint16_t updateRssi() {
+        waitForStableRssi();
```

Comment:
Another perfect place to use your Timer class 👍  (Which I really like by the way)